// mindee/src/documents/shipping_container/shippingContainerV1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShippingContainerV1 = void 0;
const document_1 = require("../document");
const fields_1 = require("../../fields");
class ShippingContainerV1 extends document_1.Document {
    constructor({ prediction, orientation = undefined, extras = undefined, inputSource = undefined, pageId = undefined, }) {
        super({
            inputSource: inputSource,
            pageId: pageId,
            orientation: orientation,
            extras: extras,
        });
        this.owner = new fields_1.Field({
            prediction: prediction.owner,
            pageId: pageId,
        });
        this.serialNumber = new fields_1.Field({
            prediction: prediction.serial_number,
            pageId: pageId,
        });
        this.sizeType = new fields_1.Field({
            prediction: prediction.size_type,
            pageId: pageId,
        });
    }
    toString() {
        const outStr = `----- Shipping Container V1 -----
Owner: ${this.owner}
Serial number: ${this.serialNumber}
Size and type: ${this.sizeType}
----------------------
`;
        return ShippingContainerV1.cleanOutString(outStr);
    }
}
exports.ShippingContainerV1 = ShippingContainerV1;
