// mindee/src/documents/fr/idCard/idCardV1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdCardV1 = void 0;
const document_1 = require("../../document");
const fields_1 = require("../../../fields");
class IdCardV1 extends document_1.Document {
    constructor({ prediction, orientation = undefined, extras = undefined, inputSource = undefined, pageId = undefined, }) {
        super({
            inputSource: inputSource,
            pageId: pageId,
            orientation: orientation,
            extras: extras,
        });
        /** The list of the names of the person. */
        this.givenNames = [];
        this.authority = new fields_1.Field({
            prediction: prediction.authority,
            pageId: pageId,
        });
        this.documentSide = new fields_1.BaseField({
            prediction: prediction.document_side,
        });
        this.idNumber = new fields_1.Field({
            prediction: prediction.id_number,
            pageId: pageId,
        });
        this.birthDate = new fields_1.DateField({
            prediction: prediction.birth_date,
            pageId: pageId,
        });
        this.expiryDate = new fields_1.DateField({
            prediction: prediction.expiry_date,
            pageId: pageId,
        });
        this.birthPlace = new fields_1.Field({
            prediction: prediction.birth_place,
            pageId: pageId,
        });
        this.gender = new fields_1.Field({
            prediction: prediction.gender,
            pageId: pageId,
        });
        this.surname = new fields_1.Field({
            prediction: prediction.surname,
            pageId: pageId,
        });
        this.mrz1 = new fields_1.Field({
            prediction: prediction.mrz1,
            pageId: pageId,
        });
        this.mrz2 = new fields_1.Field({
            prediction: prediction.mrz2,
            pageId: pageId,
        });
        prediction.given_names.map((prediction) => this.givenNames.push(new fields_1.Field({
            prediction: prediction,
            pageId: pageId,
        })));
    }
    toString() {
        const outStr = `----- FR ID Card V1 -----
Filename: ${this.filename}
Document side: ${this.documentSide}
Authority: ${this.authority}
Given names: ${this.givenNames.map((name) => name.value).join(" ")}
Surname: ${this.surname}
Gender: ${this.gender}
ID Number: ${this.idNumber}
Birth date: ${this.birthDate}
Birth place: ${this.birthPlace}
Expiry date: ${this.expiryDate}
MRZ 1: ${this.mrz1}
MRZ 2: ${this.mrz2}
----------------------
`;
        return IdCardV1.cleanOutString(outStr);
    }
}
exports.IdCardV1 = IdCardV1;
